[ADD] l10n_in: Add IRN Number Field to Journal Entry and Update UOM Code

zepa-odoo opened the pull request:
This pull request updates the Indian localization module with the following improvements:

1. **UOM Code Update:**
   - Updated the Unit of Measure (UOM) code for compliance with the latest standards.

2. **IRN Number Integration:**
   - Added the `l10n_in_irn_number` field to the `account.move` model to store Invoice Reference Numbers (IRNs).
   - Updated invoice views to display the IRN field for Indian invoices.
   - Modified `account_edi_format` to fetch and store IRNs from the GST portal.

These changes enhance compliance with Indian GST regulations and streamline the invoicing process.

**Impact:**
- Automated IRN fetching and storage.

Related-Enterprise: https://github.com/odoo/enterprise/pull/67936

Task ID: 3551097

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/175644.png)](https://mergebot.odoo.com/odoo/odoo/pull/175644)

yosuanicolaus commented on addons/account/views/account_journal_dashboard_view.xml:
```diff
@@ -309,28 +309,28 @@
+                                <widget name="account_onboarding" t-if="journal_is_empty and dashboard['onboarding']"/>
+                            </t>
+                            <t t-if="journal_type == 'purchase'">
+                                <t t-if="!journal_is_empty">
```
Since the `t-else` on the next element is a `div`, let's also refactor the only div child of this `t` element to here.

So it'll be `<div t-if="!journal_is_empty" class="d-flex"> <widget ...`

yosuanicolaus commented on addons/l10n_in/models/res_partner.py:
```diff
@@ -72,3 +72,28 @@ def check_vat_in(self, vat):
+        if partner_data:
+            partner_gid = partner_data.get('partner_gid')
+            if partner_gid:
+                partner_data = self.env['res.partner'].enrich_company(company_domain=None, partner_gid=partner_gid, vat=partner_data.get('vat'))
+                partner_data = self.env['res.partner']._iap_replace_logo(partner_data)
```
same here, no need for the `env[...]`

yosuanicolaus commented on addons/l10n_in/models/res_partner.py:
```diff
@@ -72,3 +72,28 @@ def check_vat_in(self, vat):
+                'street': partner_data.get('street'),
+                'street2': partner_data.get('street2'),
+                'city': partner_data.get('city'),
+                'state_id': partner_data.get('state_id') and partner_data.get('state_id').get('id') or False,
+                'country_id': partner_data.get('country_id') and partner_data.get('country_id').get('id') or False,
```
Were there any issue here previously? Why did you add all the ` and ...get.. or False` ?

If the `state_id` or `country_id` was not already an `id`, then you can do this

`partner_data.get('state_id', {}).get('id', False)`

But I think it should already be an `id`...

yosuanicolaus reviewed (changes requested):
Hello :wave: is this PR ready for review? anyway, here's some comments

zepa-odoo commented:
> Hello 👋 is this PR ready for review? anyway, here's some comments

Yes, it’s ready for review! 
I’ve addressed your comments as well. 
Thank you for your feedback!

yosuanicolaus commented on addons/l10n_in/models/res_partner.py:
```diff
@@ -72,3 +72,28 @@ def check_vat_in(self, vat):
+                'street': partner_data.get('street'),
+                'street2': partner_data.get('street2'),
+                'city': partner_data.get('city'),
+                'state_id': partner_data.get('state_id', False),
+                'country_id': partner_data.get('country_id', False),
```
Do we need the `False` default value here? Isn't `None` from `get` enough if we're just using this dict in `.create(...)` ?

zepa-odoo replied:
I will keep my the code back:

```python
'state_id': partner_data.get('state_id') and partner_data.get('state_id').get('id') or False,
```

The `read_by_vat` method returns `country_id` and `state_id` as dictionaries that include both `id` and `display_name`. Therefore, it's necessary to set the value to `False` if no data is retrieved.

For reference, you can check the code here where the values are set: Code Reference https://github.com/odoo/odoo/blob/5ad3315b260206be564c6ea08328d237bc656f5f/addons/partner_autocomplete/models/res_partner.py#L56

yosuanicolaus replied:
I see, then let's improve it with my first suggestion then.
`'state_id': partner_data.get('state_id', {}).get('id', False)`

proof in python:
```py
>>> a = {'c': {'id': 123}}
>>> b = {}
>>> a.get('c', {}).get('id', False)
123
>>> b.get('c', {}).get('id', False)
False

zepa-odoo replied:
Hello @yosuanicolaus ,
I have made the changes.

yosuanicolaus reviewed (changes requested):
Almost good. Don't forget to squash your commit at the end

jco-odoo commented:
@robodoo r+

robodoo commented:
@jco-odoo you may want to rebuild or fix this PR as it has failed CI.
